README.md
MaRTEnoid for MaRTE OS
=================

MaRTEnoid was programmed as an example for MaRTE OS of how
to use its vga driver in C with double buffering. 

The code is not optimized and can be improved in some areas 
like collision detection and ball physics, but the goal of 
this program was making a quick, working and hopefully fun 
example, not making the perfect game implementation :)

This code is distributed for educational purposes only.
License: GPL v2

This is the original version I made in 2009.

Video: https://www.youtube.com/watch?v=FWex5FDVWXA

/////////////////////////////////////////////////////

0- INSTALLATION:

A makefile is provided to easily generate the program, however, 
you may need to change the path to the svga driver to the one 
corresponding to your MaRTE OS installation.

1- BASIC CONTROLS:

left arrow or "a" = move left
right arrow or "d" = move right

2- GAMEPLAY:

This is a little tribute to the classic Arkanoid, so most things 
are similar. 
Silver (dark gray) bricks need to be hit two times to be destroyed,
Golden ones can't be destroyed.

There're 7 levels.

ChangeLog
2009/06/03: added code to print bitmap fonts
2009/05/30: first public version


maps.h
/*---------------------------------------------------------------------------
--                               MaRTEnoid                                 -- 
-----------------------------------------------------------------------------
--                                                                         --
--  MaRTEnoid is an example game for MaRTE OS.                             --
--  This code is distributed for educational purposes only.                --
--                                                                         --
--  author:   Alvaro Garcia Cuesta                                         --
--  email:    alvaro@binarynonsense.com                                    --
--  website:  www.binarynonsense.com                                       --
--                                                                         --
--  file: maps.h                                                           --
--                                                                         --
--  this file contains the level maps used in the game                     --
-----------------------------------------------------------------------------
--                               License                                  -- 
-----------------------------------------------------------------------------
--                                                                         --
-- MaRTEnoid is free software; you can redistribute it and/or modify it    --
-- under the terms of the GNU General Public License version 2 as          -- 
-- published by the Free Software Foundation.                              --
-- See COPYING file for more info about the license                        --
--                                                                         --
-----------------------------------------------------------------------------
--                        last update: 28 May 09                           --
---------------------------------------------------------------------------*/

#define numBricksWidth 13
#define numBricksHeight 26

/*
8 normal: 50pts+10
grey orange bluelight green red blue pink yellow
1        2     3         4   5   6    7     8
2 special:
greydark gold dark
9-11      12
*/

int map_19[numBricksWidth*numBricksHeight]=
{
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,10,10,10,10,10,10,10,0,0,0,
    0,0,4,5,6,7,10,7,6,5,4,0,0,
    0,0,4,5,6,7,10,7,6,5,4,0,0,
    0,0,4,5,6,7,10,7,6,5,4,0,0,
    0,0,4,5,6,7,0,7,6,5,4,0,0,
    0,0,4,5,6,7,10,7,6,5,4,0,0,
    0,0,4,5,6,7,10,7,6,5,4,0,0,
    0,0,4,5,6,7,10,7,6,5,4,0,0,
    0,0,0,10,10,10,10,10,10,10,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0
};

int map_7[numBricksWidth*numBricksHeight]=
{
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,8,8,7,0,0,0,0,0,
    0,0,0,0,8,8,7,7,6,0,0,0,0,
    0,0,0,8,8,7,7,6,6,5,0,0,0,
    0,0,0,8,7,7,6,6,5,5,0,0,0,
    0,0,8,7,7,6,6,5,5,4,4,0,0,
    0,0,7,7,6,6,5,5,4,4,3,0,0,
    0,0,7,6,6,5,5,4,4,3,3,0,0,
    0,0,6,6,5,5,4,4,3,3,2,0,0,
    0,0,6,5,5,4,4,3,3,2,2,0,0,
    0,0,5,5,4,4,3,3,2,2,1,0,0,
    0,0,0,4,4,3,3,2,2,1,0,0,0,
    0,0,0,4,3,3,2,2,1,1,0,0,0,
    0,0,0,0,3,2,2,1,1,0,0,0,0,
    0,0,0,0,0,2,1,1,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0
};

int map_5[numBricksWidth*numBricksHeight]=
{
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,8,0,0,0,0,0,8,0,0,0,
    0,0,0,8,0,0,0,0,0,8,0,0,0,
    0,0,0,0,8,0,0,0,8,0,0,0,0,
    0,0,0,0,8,0,0,0,8,0,0,0,0,
    0,0,0,9,9,9,9,9,9,9,0,0,0,
    0,0,0,9,9,9,9,9,9,9,0,0,0,
    0,0,9,9,5,9,9,9,5,9,9,0,0,
    0,0,9,9,5,9,9,9,5,9,9,0,0,
    0,9,9,9,9,9,9,9,9,9,9,9,0,
    0,9,9,9,9,9,9,9,9,9,9,9,0,
    0,9,9,9,9,9,9,9,9,9,9,9,0,
    0,9,0,9,9,9,9,9,9,9,0,9,0,
    0,9,0,9,0,0,0,0,0,9,0,9,0,
    0,9,0,9,0,0,0,0,0,9,0,9,0,
    0,0,0,0,9,9,0,9,9,0,0,0,0,
    0,0,0,0,9,9,0,9,9,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0
};

int map_1[numBricksWidth*numBricksHeight]=
{
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    4,4,4,4,4,4,4,4,4,4,4,4,4,
    3,3,3,3,3,3,3,3,3,3,3,3,3,
    2,2,2,2,2,2,2,2,2,2,2,2,2,
    1,1,1,1,1,1,1,1,1,1,1,1,1,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0
};

int map_2[numBricksWidth*numBricksHeight]=
{
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    1,0,0,0,0,0,0,0,0,0,0,0,0,
    1,2,0,0,0,0,0,0,0,0,0,0,0,
    1,2,3,0,0,0,0,0,0,0,0,0,0,
    1,2,3,1,0,0,0,0,0,0,0,0,0,
    1,2,3,1,2,0,0,0,0,0,0,0,0,
    1,2,3,1,2,3,0,0,0,0,0,0,0,
    1,2,3,1,2,3,1,0,0,0,0,0,0,
    1,2,3,1,2,3,1,2,0,0,0,0,0,
    1,2,3,1,2,3,1,2,3,0,0,0,0,
    1,2,3,1,2,3,1,2,3,1,0,0,0,
    1,2,3,1,2,3,1,2,3,1,2,0,0,
    1,2,3,1,2,3,1,2,3,1,2,3,0,
    9,9,9,9,9,9,9,9,9,9,9,9,9,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0
};


int map_marte[numBricksWidth*numBricksHeight]=
{
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    4,0,0,0,4,0,0,0,3,3,3,3,0,
    4,4,0,4,4,2,2,2,3,0,0,0,3,
    4,0,4,0,4,2,0,2,3,3,3,3,3,
    4,0,0,0,4,2,2,2,3,0,0,3,0,
    4,0,0,0,4,2,0,2,3,0,0,0,3,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,6,6,6,0,5,5,5,0,0,0,
    0,0,0,0,6,0,0,5,0,0,0,0,0,
    0,0,0,0,6,0,0,5,5,5,0,0,0,
    0,0,0,0,6,0,0,5,0,0,0,0,0,
    0,0,0,0,6,0,0,5,5,5,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0
};

int map_4[numBricksWidth*numBricksHeight]=
{
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,2,3,5,9,6,0,8,1,2,3,5,0,
    0,3,5,9,6,7,0,1,2,3,5,9,0,
    0,5,9,6,7,8,0,2,3,5,9,6,0,
    0,9,6,7,8,1,0,3,5,9,6,5,0,
    0,6,7,8,1,2,0,5,9,6,5,5,0,
    0,7,8,1,2,3,0,9,6,5,5,3,0,
    0,8,1,2,3,5,0,6,5,5,3,2,0,
    0,1,2,3,5,9,0,5,5,3,2,5,0,
    0,2,3,5,9,6,0,5,3,2,5,2,0,
    0,3,5,9,6,5,0,3,2,5,2,6,0,
    0,5,9,6,5,1,0,2,5,2,6,7,0,
    0,9,6,5,1,3,0,5,2,6,7,10,0,
    0,6,5,1,3,2,0,2,6,7,10,10,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0,
    0,0,0,0,0,0,0,0,0,0,0,0,0
};

///////blank map template://////////
// int map_1[numBricksWidth*numBricksHeight]=
// {
//     0,0,0,0,0,0,0,0,0,0,0,0,0,
//     0,0,0,0,0,0,0,0,0,0,0,0,0,
//     0,0,0,0,0,0,0,0,0,0,0,0,0,
//     0,0,0,0,0,0,0,0,0,0,0,0,0,
//     0,0,0,0,0,0,0,0,0,0,0,0,0,
//     0,0,0,0,0,0,0,0,0,0,0,0,0,
//     0,0,0,0,0,0,0,0,0,0,0,0,0,
//     0,0,0,0,0,0,0,0,0,0,0,0,0,
//     0,0,0,0,0,0,0,0,0,0,0,0,0,
//     0,0,0,0,0,0,0,0,0,0,0,0,0,
//     0,0,0,0,0,0,0,0,0,0,0,0,0,
//     0,0,0,0,0,0,0,0,0,0,0,0,0,
//     0,0,0,0,0,0,0,0,0,0,0,0,0,
//     0,0,0,0,0,0,0,0,0,0,0,0,0,
//     0,0,0,0,0,0,0,0,0,0,0,0,0,
//     0,0,0,0,0,0,0,0,0,0,0,0,0,
//     0,0,0,0,0,0,0,0,0,0,0,0,0,
//     0,0,0,0,0,0,0,0,0,0,0,0,0,
//     0,0,0,0,0,0,0,0,0,0,0,0,0,
//     0,0,0,0,0,0,0,0,0,0,0,0,0,
//     0,0,0,0,0,0,0,0,0,0,0,0,0,
//     0,0,0,0,0,0,0,0,0,0,0,0,0,
//     0,0,0,0,0,0,0,0,0,0,0,0,0,
//     0,0,0,0,0,0,0,0,0,0,0,0,0,
//     0,0,0,0,0,0,0,0,0,0,0,0,0,
//     0,0,0,0,0,0,0,0,0,0,0,0,0
// };


fonts.c
/*---------------------------------------------------------------------------
--                               MaRTEnoid                                 -- 
-----------------------------------------------------------------------------
--                                                                         --
--  MaRTEnoid is an example game for MaRTE OS.                             --
--  This code is distributed for educational purposes only.                --
--                                                                         --
--  author:   Alvaro Garcia Cuesta                                         --
--  email:    alvaro@binarynonsense.com                                    --
--  website:  www.binarynonsense.com                                       --
--                                                                         --
--  file: fonts.c                                                          --
--                                                                         --
--  this file contains the functions used to print strings                 --
--  using a bitmap font                                                    --
-----------------------------------------------------------------------------
--                               License                                  -- 
-----------------------------------------------------------------------------
--                                                                         --
-- MaRTEnoid is free software; you can redistribute it and/or modify it    --
-- under the terms of the GNU General Public License version 2 as          -- 
-- published by the Free Software Foundation.                              --
-- See COPYING file for more info about the license                        --
--                                                                         --
-----------------------------------------------------------------------------
--                        last update: 02 Jun 09                           --
---------------------------------------------------------------------------*/

#include <string.h>
#include "vga.h"
#include "font.h"
#include "fonts.h"

int *theFont=font_img;
int screenWidth = 320;
int fontWidth=8;
int fontHeight=16;
int fontFileWidth=128;


void bmpfont_print(char *theString, int x, int y,unsigned char *backBuffer){
    int length;
    length = strlen(theString);
    
    int charPosition,charValue,fontPositionX,fontPositionY,color;
    int xf,yf,initX,initY;
    for(charPosition=0;charPosition<length;charPosition++){
        charValue=theString[charPosition];
        if(charValue<128){
            fontPositionY=charValue/16;
            fontPositionX=charValue-(fontPositionY*16);
            initX=fontPositionX*fontWidth;
            initY=fontPositionY*fontHeight;
            for(xf=0;xf<fontWidth;xf++){
                for(yf=0;yf<fontHeight;yf++){
                    color=theFont[fontFileWidth*(yf+initY)+(xf+initX)];
                    if(color!=0){
                        backBuffer[screenWidth*(y+yf)+(x+xf+charPosition*fontWidth)]=224;//white
                    }
                    //ONLY FOR MARTENOID
                    else{
                        backBuffer[screenWidth*(y+yf)+(x+xf+charPosition*fontWidth)]=0;//black
                    }
                }
            }
            
        }
        
    }
}


fonts.h
/*---------------------------------------------------------------------------
--                               MaRTEnoid                                 -- 
-----------------------------------------------------------------------------
--                                                                         --
--  MaRTEnoid is an example game for MaRTE OS.                             --
--  This code is distributed for educational purposes only.                --
--                                                                         --
--  author:   Alvaro Garcia Cuesta                                         --
--  email:    alvaro@binarynonsense.com                                    --
--  website:  www.binarynonsense.com                                       --
--                                                                         --
--  file: fonts.h                                                          --
--                                                                         --
--  this file contains the functions used to print strings                 --
--  using a bitmap font                                                    --
-----------------------------------------------------------------------------
--                               License                                  -- 
-----------------------------------------------------------------------------
--                                                                         --
-- MaRTEnoid is free software; you can redistribute it and/or modify it    --
-- under the terms of the GNU General Public License version 2 as          -- 
-- published by the Free Software Foundation.                              --
-- See COPYING file for more info about the license                        --
--                                                                         --
-----------------------------------------------------------------------------
--                        last update: 02 Jun 09                           --
---------------------------------------------------------------------------*/

void bmpfont_print(char *theString, int x, int y,unsigned char *backBuffer);


main.h
/*---------------------------------------------------------------------------
--                               MaRTEnoid                                 -- 
-----------------------------------------------------------------------------
--                                                                         --
--  MaRTEnoid is an example game for MaRTE OS.                             --
--  This code is distributed for educational purposes only.                --
--                                                                         --
--  author:   Alvaro Garcia Cuesta                                         --
--  email:    alvaro@binarynonsense.com                                    --
--  website:  www.binarynonsense.com                                       --
--                                                                         --
--  file: main.h                                                           --
--                                                                         --
--  this file is the header for main.c                                     --
-----------------------------------------------------------------------------
--                               License                                  -- 
-----------------------------------------------------------------------------
--                                                                         --
-- MaRTEnoid is free software; you can redistribute it and/or modify it    --
-- under the terms of the GNU General Public License version 2 as          -- 
-- published by the Free Software Foundation.                              --
-- See COPYING file for more info about the license                        --
--                                                                         --
-----------------------------------------------------------------------------
--                        last update: 03 Jun 09                           --
---------------------------------------------------------------------------*/

#include "vga.h"
#include <assert.h> // assert
#include <time.h>   // timespec, nanosleep
#include <math.h>   // sin
#include <misc/timespec_operations.h>  // double_to_timespec
#include <misc/console_management.h> // reset_blocking.. set_raw_mode
#include <string.h> //memset

#define PCI_DEVICE_ID_S3_TRIO64V2 35073

#include "images.h"
#include "maps.h"
#include "fonts.h"

//VARIABLES///////////////////

int gameOver;
double period;
struct timespec period_ts;
int screenWidth,screenHeight;

unsigned char *backBuffer;

#define brickWidth 10
#define brickHeight 5
#define numBricksWidth 13
#define numBricksHeight 26

//ball velocity can be adjusted setting the following variables:
//ballFramesChange, ball_dY, and the period
int frameNumber=0;
int ballFramesChange=0;

int actualMap=0;
int numMaps=7;
int *map[7];
int mapBricks=0;

int *bricksImages[10];

int areaX=0;
int areaY=0;
int areaWidth=numBricksWidth*brickWidth;
int areaBricksHeight=numBricksHeight*brickHeight;
int areaHeight;

int playerWidth=20;
int playerHeight=5;
int playerX=0;
int playerY=0;
int playerVelocity=6;//6 for real pc, 6 for qemu

int ballWidth=5;
int ballHeight=4;
int ballX=0;
int ballY=0;
int ball_dX=0;
int ball_dY=1;//1 for real pc, 4 for qemu

int backgroundBorderWidth=136;
int backgroundBorderHeight=171;
int backgroundBorderPosX,backgroundBorderPosY;
int backgroundWidth=130;
int backgroundHeight=165;
int backgroundPosX,backgroundPosY;

char text[50];
int lives=3;
int message=0;//0 is gameover 1 is win

//FUNCTIONS///////////////////

void init();

void initPalette();

void draw_border();

void draw_gameover(int message);

void draw();

void move();

void checkCollisions();

int numBricksMap(int mapNumber);

void nextLevel();

void blit();

int gameLoop();

